File: AppBox/Common/UserManager/UserData.m
//
//  UserData.m
//  AppBox
//
//  Created by Vineet Choudhary on 12/12/16.
//  Copyright © 2016 Developer Insider. All rights reserved.
//

#import "UserData.h"


@implementation UserData

//MARK: - Logged In User
+(BOOL)isLoggedIn {
	return [DBClientsManager authorizedClients] && ([DBClientsManager authorizedClients].count > 0);
}

#define LoggedInUserEmail @"LoggedInUserEmail"
+(NSString *)loggedInUserEmail {
	NSString *userEmail = [[NSUserDefaults standardUserDefaults] stringForKey:LoggedInUserEmail];
	return userEmail == nil ? abEmptyString : userEmail;
}

+(void)setLoggedInUserEmail:(NSString *)loggedInUserEmail {
	[[NSUserDefaults standardUserDefaults] setValue:loggedInUserEmail forKey:LoggedInUserEmail];
	[[NSUserDefaults standardUserDefaults] synchronize];
}

#define LoggedInUserDisplayName @"LoggedInUserDisplayName"
+(NSString *)loggedInUserDisplayName {
	NSString *displayName = [[NSUserDefaults standardUserDefaults] stringForKey:LoggedInUserDisplayName];
	return displayName == nil ? abEmptyString : displayName;
}

+(void)setLoggedInUserDisplayName:(NSString *)loggedInUserDisplayName {
	[[NSUserDefaults standardUserDefaults] setValue:loggedInUserDisplayName forKey:LoggedInUserDisplayName];
	[[NSUserDefaults standardUserDefaults] synchronize];
}

//MARK: - Dropbox -

#define DropboxUsedSpace @"DropboxUsedSpace"

+(NSNumber *)dropboxUsedSpace{
	return @([[NSUserDefaults standardUserDefaults] doubleForKey:DropboxUsedSpace]);
}

+(void)setDropboxUsedSpace:(NSNumber *)usedSpace{
	[[NSUserDefaults standardUserDefaults] setInteger:usedSpace.integerValue forKey:DropboxUsedSpace];
	[[NSUserDefaults standardUserDefaults] synchronize];
}

//MARK: - Preferences...
//MARK: - Email Releated -

#define UserEmail @"UserEmail"

+(NSString *)userEmail{
    NSString *userEmail = [[NSUserDefaults standardUserDefaults] stringForKey:UserEmail];
    return userEmail == nil ? abEmptyString : userEmail;
}

+(void)setUserEmail:(NSString *)userEmail{
    [[NSUserDefaults standardUserDefaults] setValue:userEmail forKey:UserEmail];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

#define UserMessage @"UserMessage"

+(NSString *)userMessage{
    NSString *userMessage = [[NSUserDefaults standardUserDefaults] stringForKey:UserMessage];
    return userMessage == nil ? abEmptyString : userMessage;
}

+(void)setUserMessage:(NSString *)userMessage{
    [[NSUserDefaults standardUserDefaults] setValue:userMessage forKey:UserMessage];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

#define UserSlackChannel @"UserSlackChannel"

+(NSString *)userSlackChannel {
    NSString *userSlackChannel = [[NSUserDefaults standardUserDefaults] stringForKey:UserSlackChannel];
    return userSlackChannel == nil ? abEmptyString : userSlackChannel;
}

+(void)setUserSlackChannel:(NSString *)slackChannel {
    [[NSUserDefaults standardUserDefaults] setValue:slackChannel forKey:UserSlackChannel];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

#define UserHangoutChatWebHook @"UserHangoutChatWebHook"

+(NSString *)userHangoutChatWebHook {
    NSString *userHangoutChatWebHook = [[NSUserDefaults standardUserDefaults] stringForKey:UserHangoutChatWebHook];
    return userHangoutChatWebHook == nil ? abEmptyString : userHangoutChatWebHook;
}

+(void)setUserHangoutChatWebHook:(NSString *)slackChannel {
    [[NSUserDefaults standardUserDefaults] setValue:slackChannel forKey:UserHangoutChatWebHook];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

#define UserMicrosoftTeamWebHook @"UserMicrosoftTeamWebHook"

+(NSString *)userMicrosoftTeamWebHook {
    NSString *userMicrosoftTeamWebHook = [[NSUserDefaults standardUserDefaults] stringForKey:UserMicrosoftTeamWebHook];
    return userMicrosoftTeamWebHook == nil ? abEmptyString : userMicrosoftTeamWebHook;
}

+(void)setUserMicrosoftTeamWebHook:(NSString *)slackChannel {
    [[NSUserDefaults standardUserDefaults] setValue:slackChannel forKey:UserMicrosoftTeamWebHook];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

#define UserSlackMessage @"UserSlackMessage"

+(NSString *)userSlackMessage {
    NSString *userSlackMessage = [[NSUserDefaults standardUserDefaults] stringForKey:UserSlackMessage];
    return userSlackMessage == nil ? abEmptyString : userSlackMessage;
}

+(void)setUserSlackMessage:(NSString *)slackMessage {
    [[NSUserDefaults standardUserDefaults] setValue:slackMessage forKey:UserSlackMessage];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

#define DropboxAvailableSpace @"DropboxAvailableSpace"

+(NSNumber *)dropboxAvailableSpace{
    return @([[NSUserDefaults standardUserDefaults] doubleForKey:DropboxAvailableSpace]);
}

+(void)setDropboxAvailableSpace:(NSNumber *)availableSpace{
    [[NSUserDefaults standardUserDefaults] setInteger:availableSpace.integerValue forKey:DropboxAvailableSpace];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

//MARK: - AppBox Installation Page Settings -
#define DonwloadIPAEnable @"DonwloadIPAEnable"
+(BOOL)downloadIPAEnable{
    return [[NSUserDefaults standardUserDefaults] boolForKey:DonwloadIPAEnable];
}

+(void)setDownloadIPAEnable:(BOOL)downloadIPA{
    [[NSUserDefaults standardUserDefaults] setBool:downloadIPA forKey:DonwloadIPAEnable];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

#define MoreDetailsEnable @"MoreDetailsEnable"
+(BOOL)moreDetailsEnable{
    return [[NSUserDefaults standardUserDefaults] boolForKey:MoreDetailsEnable];
}

+(void)setMoreDetailsEnable:(BOOL)moreDetails{
    [[NSUserDefaults standardUserDefaults] setBool:moreDetails forKey:MoreDetailsEnable];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

#define ShowPreviousVersions @"ShowPreviousVersions"
+(BOOL)showPreviousVersions{
    return [[NSUserDefaults standardUserDefaults] boolForKey:ShowPreviousVersions];
}

+(void)setShowPreviousVersions:(BOOL)previousVersion{
    [[NSUserDefaults standardUserDefaults] setBool:previousVersion forKey:ShowPreviousVersions];
    [[NSUserDefaults standardUserDefaults] synchronize];
}


//MARK: - App Settings -
#define AppSettingIsFirstTime @"AppSettingIsFirstTime"
+(BOOL)isFirstTime{
    return ![[NSUserDefaults standardUserDefaults] boolForKey:AppSettingIsFirstTime];
}

+(void)setIsFirstTime:(BOOL)isFirstTime{
    [[NSUserDefaults standardUserDefaults] setBool:isFirstTime forKey:AppSettingIsFirstTime];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

+(NSString *)appSettingIsFirstTimeAfterUpdate{
    NSString *versionString = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"CFBundleShortVersionString"];
    return [NSString stringWithFormat:@"AppSettingIsFirstTimeAfterUpdate%@", versionString];
}

+(BOOL)isFirstTimeAfterUpdate{
    return ![[NSUserDefaults standardUserDefaults] boolForKey:[self appSettingIsFirstTimeAfterUpdate]];
}

+(void)setIsFirstTimeAfterUpdate:(BOOL)isFirstTime{
    [[NSUserDefaults standardUserDefaults] setBool:isFirstTime forKey:[self appSettingIsFirstTimeAfterUpdate]];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

//MARK: - Chunk Size -
#define UploadChunkSize @"UploadChunkSize"
+(NSInteger)uploadChunkSize{
    NSInteger chunkSize = [[NSUserDefaults standardUserDefaults] integerForKey:UploadChunkSize];
    return chunkSize > 0 ? chunkSize : 100;
}

+(void)setUploadChunkSize:(NSInteger)chunkSize{
    [[NSUserDefaults standardUserDefaults] setInteger:chunkSize forKey:UploadChunkSize];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

//MARK: - CI Settings -
#define DebugLogEnable @"DebugLogEnable"
+(BOOL)debugLog{
    return [[NSUserDefaults standardUserDefaults] boolForKey:DebugLogEnable];
}

+(void)setEnableDebugLog:(BOOL)debugLog{
    [[NSUserDefaults standardUserDefaults] setBool:debugLog forKey:DebugLogEnable];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

#define UpdateAlertEnable @"UpdateAlertEnable"
+(BOOL)updateAlertEnable{
    return [[NSUserDefaults standardUserDefaults] boolForKey:UpdateAlertEnable];
}

+(void)setUpdateAlertEnable:(BOOL)updateAlert{
    [[NSUserDefaults standardUserDefaults] setBool:updateAlert forKey:UpdateAlertEnable];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

@end

File: AppBox/Common/UserManager/UserData.swift
//
//  UserData.swift
//  AppBox

import Foundation
import AppBoxCore

public final class UserData: NSObject {

    private static var defaults: UserDefaults { .standard }

    // MARK: - Logged In User

	public class func isLoggedIn() -> Bool {
		DropboxSession.isAuthorized
	}

	public class func loggedInUserEmail() -> String {
		defaults.string(forKey: "LoggedInUserEmail") ?? ""
	}

	public class func setLoggedInUserEmail(_ value: String?) {
		defaults.set(value, forKey: "LoggedInUserEmail")
	}

	public class func loggedInUserDisplayName() -> String {
		defaults.string(forKey: "LoggedInUserDisplayName") ?? ""
	}

	public class func setLoggedInUserDisplayName(_ value: String?) {
		defaults.set(value, forKey: "LoggedInUserDisplayName")
	}

    // MARK: - Dropbox space

	public class func dropboxUsedSpace() -> NSNumber {
		NSNumber(value: defaults.double(forKey: "DropboxUsedSpace"))
	}

	public class func setDropboxUsedSpace(_ usedSpace: NSNumber) {
		defaults.set(usedSpace.intValue, forKey: "DropboxUsedSpace")
	}

	public class func dropboxAvailableSpace() -> NSNumber {
		NSNumber(value: defaults.double(forKey: "DropboxAvailableSpace"))
	}

	public class func setDropboxAvailableSpace(_ availableSpace: NSNumber) {
		defaults.set(availableSpace.intValue, forKey: "DropboxAvailableSpace")
	}

    // MARK: - Email / message

	public class func userEmail() -> String {
		defaults.string(forKey: "UserEmail") ?? ""
	}

	public class func setUserEmail(_ value: String?) {
		defaults.set(value, forKey: "UserEmail")
	}

	public class func userMessage() -> String {
		defaults.string(forKey: "UserMessage") ?? ""
	}

	public class func setUserMessage(_ value: String?) {
		defaults.set(value, forKey: "UserMessage")
	}

    // MARK: - Webhooks (secrets -> Keychain via Core WebhookSecrets; S3)

	public class func userSlackChannel() -> String {
		WebhookSecrets.shared.slackWebhook ?? ""
	}

	public class func setUserSlackChannel(_ slackChannel: String?) {
		WebhookSecrets.shared.slackWebhook = slackChannel
	}

	public class func userMicrosoftTeamWebHook() -> String {
		WebhookSecrets.shared.teamsWebhook ?? ""
	}

	public class func setUserMicrosoftTeamWebHook(_ value: String?) {
		WebhookSecrets.shared.teamsWebhook = value
	}

    // MARK: - Install-page settings

	public class func downloadIPAEnable() -> Bool {
		defaults.bool(forKey: "DonwloadIPAEnable")
	}

	public class func setDownloadIPAEnable(_ downloadIPA: Bool) {
		defaults.set(downloadIPA, forKey: "DonwloadIPAEnable")
	}

	public class func moreDetailsEnable() -> Bool {
		defaults.bool(forKey: "MoreDetailsEnable")
	}

	public class func setMoreDetailsEnable(_ moreDetails: Bool) {
		defaults.set(moreDetails, forKey: "MoreDetailsEnable")
	}

	public class func showPreviousVersions() -> Bool {
		defaults.bool(forKey: "ShowPreviousVersions")
	}

	public class func setShowPreviousVersions(_ previousVersion: Bool) {
		defaults.set(previousVersion, forKey: "ShowPreviousVersions")
	}

    // MARK: - App user check

	public class func isFirstTime() -> Bool {
		!defaults.bool(forKey: "AppSettingIsFirstTime")
	}

	public class func setIsFirstTime(_ isFirstTime: Bool) {
		defaults.set(isFirstTime, forKey: "AppSettingIsFirstTime")
	}

    private static var currentAppVersion: String {
        Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""
    }

	public class func lastLaunchedVersion() -> String {
		defaults.string(forKey: "AppBoxLastLaunchedVersion") ?? ""
	}

	public class func isFirstTimeAfterUpdate() -> Bool {
		let previous = lastLaunchedVersion()
		guard !previous.isEmpty else {
			return false
		}
		return VersionCompare.isUpdateAvailable(latest: currentAppVersion, current: previous)
	}

	public class func recordLaunchedVersion() {
		defaults.set(currentAppVersion, forKey: "AppBoxLastLaunchedVersion")
	}

    // MARK: - Chunk size

    public class func uploadChunkSize() -> Int {
        let chunkSize = defaults.integer(forKey: "UploadChunkSize")
        return chunkSize > 0 ? chunkSize : 100
    }

	public class func setUploadChunkSize(_ chunkSize: Int) {
		defaults.set(chunkSize, forKey: "UploadChunkSize")
	}

    // MARK: - General settings

	public class func debugLog() -> Bool {
		defaults.bool(forKey: "DebugLogEnable")
	}

	public class func setEnableDebugLog(_ debugLog: Bool) {
		defaults.set(debugLog, forKey: "DebugLogEnable")
	}

	public class func updateAlertEnable() -> Bool {
		defaults.bool(forKey: "UpdateAlertEnable")
	}

	public class func setUpdateAlertEnable(_ updateAlert: Bool) {
		defaults.set(updateAlert, forKey: "UpdateAlertEnable")
	}

    // MARK: - CLI

	public class func cliVersion() -> String {
		defaults.string(forKey: "CLIVersion") ?? ""
	}

	public class func setCLIVersion(_ cliVersion: String?) {
		defaults.set(cliVersion, forKey: "CLIVersion")
	}
}
